keras_nlp/src/models/video_swin/video_swin_layers.py

```diff
@@ -0,0 +1,1015 @@
+        shift_size: A tuple of ints of size 3 (optional) representing the window
+            shifting size (depth, height, width) for shifted window processing
+            used in Swin Transformer. If not provided, only window size is computed.
+
+    Returns:
+        A tuple or a pair of tuples:
+            - If `shift_size` is None, returns a single tuple representing the adjusted
+            window size that may be smaller than the provided `window_size` to ensure
+            it doesn't exceed the input size along any dimension.
+            - If `shift_size` is provided, returns a pair of tuples. The first tuple
+            represents the adjusted window size, and the second tuple represents the
+            adjusted shift size. The adjustments ensure both window size and shift size
+            do not exceed the corresponding dimensions in the input data.
+    """
+
+    use_window_size = list(window_size)
+
+    if shift_size is not None:
+        use_shift_size = list(shift_size)
+
+    for i in range(len(x_size)):
+        if x_size[i] <= window_size[i]:
+            use_window_size[i] = x_size[i]
+            if shift_size is not None:
+                use_shift_size[i] = 0
+
+    if shift_size is None:
+        return tuple(use_window_size)
+    else:
+        return tuple(use_window_size), tuple(use_shift_size)
+
+
+def compute_mask(depth, height, width, window_size, shift_size):
+    """Computes an attention mask for a sliding window self-attention mechanism
+    used in Video Swin Transformer.
+
+    This function creates a mask to indicate which windows can attend to each other
+    during the self-attention operation. It considers non-overlapping and potentially
+    shifted windows based on the provided window size and shift size.
+
+    Args:
+        depth (int): Depth (number of frames) of the input video.
+        height (int): Height of the video frames.
+        width (int): Width of the video frames.
+        window_size (tuple[int]): Size of the sliding window in each dimension
+            (depth, height, width).
+        shift_size (tuple[int]): Size of the shifting step in each dimension
+            (depth, height, width).
+
+    Returns:
+        A tensor of shape (batch_size, num_windows, num_windows), where:
+            - batch_size: Assumed to be 1 in this function.
+            - num_windows: Total number of windows covering the entire input based on
+                the formula:
+                    (depth - window_size[0]) // shift_size[0] + 1) *
+                    (height - window_size[1]) // shift_size[1] + 1) *
+                    (width - window_size[2]) // shift_size[2] + 1)
+        Each element (attn_mask[i, j]) represents the attention weight between
+        window i and window j. A value of -100.0 indicates high negative attention
+        (preventing information flow), 0.0 indicates no mask effect.
+    """
+
+    img_mask = np.zeros((1, depth, height, width, 1))
```

Comment:
what's the overall structure of this mask? we likely want to find a way to rewrite this whole block using `ops` and not `numpy` and for loops.